### ISSUE TITLE
Ability to add webauthn/passkey authentication for the web interface

### ISSUE THREAD

**Issue by ThatOneCalculator** [cockpit-project/cockpit#20389]
Since right now all it takes to get in through the web interface is a username & password, being able to enforce the use of something like a Yubikey (like I do with my ssh keys) would be great.

**Comment by jelly** [comment]
We use PAM for authentication so if you configure totp with Yubikey you should be able to get 2fa in Cockpit. This is a duplicate of https://github.com/cockpit-project/cockpit/issues/15860

**Comment by ThatOneCalculator** [comment]
2FA/TOTP with a YubiKey isn't the same as using Webauthn with a Yubikey.

**Comment by jelly** [comment]
Overlooked that, but regardless implementing passkey/webauthn is a lot of work afaik there are no off the shelve solutions for this.

**Comment by ThatOneCalculator** [comment]
As for the backend, webauthn data is commonly sent as a JSON Web Token (JTW), and most modern backend web frameworks have a single-line solution to verify JWTs. https://jwt.io/libraries

**Comment by JanZerebecki** [comment]
duplicate of https://github.com/cockpit-project/cockpit/issues/18566

**Comment by Venefilyn** [comment]
This would indeed be useful and powerful. But I feel like the blocker here is that there aren't any PAM modules we can use right? I can see https://github.com/Yubico/pam-u2f but I'm unsure how one would connect that with WebAuthn flow

**Comment by JanZerebecki** [comment]
Kanidm has a pam module that should work for this. It should be possible currently to configure that so it works. To make it so it could work out of the box maybe we could make cockpit act as reverse proxy for kanidm.

**Comment by Venefilyn** [comment]
@JanZerebecki Seems like that would be vendor-locking sorta deal if we reverse proxy for kanidm. I did look into their PAM module but it starts off with them saying it has no support for SELinux

**Comment by JanZerebecki** [comment]
Maybe file an issue asking if that is outdated or they plan to implement it? I see code interacting with selinux.

There is no reason we can't reverse proxy multiple. I also see no reason why the needed integration can not be made an open protocol, but I would not put the burden on whoever implements this first. We want to make this work out of the box with just installing the cockpit package, thus zero configuration and not cause any other disturbances on the system, so I think ease of use is more important than not depending on specific functionality.

Another implementation is sssd, but that is harder to modify.

**Comment by mguinness** [comment]
You should be able to add WebAuthn using a combination of [Proxying Cockpit over NGINX](https://github.com/cockpit-project/cockpit/wiki/Proxying-Cockpit-over-NGINX) and [NGINX + WebAuthn for your small scale web applications](https://github.com/newhouseb/nginxwebauthn) which does [Authentication Based on Subrequest Result](https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-subrequest-authentication/).

**Comment by Venefilyn** [comment]
We recently had DevConf CZ in Brno and I got to talking about this with a few people. From the convo and now reading the docs surrounding PAM and sssd I do think there is a reasonable approach here that wouldn't be too complex..

sssd itself has support for FIDO2/WebAuthn and it does seem like something we can utilize. An LDAP server is likely to be a requirement. I expect we can start to login with passkeys and then tackle the key registation later. As long as the keys are registered I think it'll all be fine.. hm

However, I'm not going to commit to anything and I am also not sure it would even work or what edge-cases there are. But I will experiment with it soon(tm) and would just like to share some details about this

**Comment by Venefilyn** [comment]
Update since last post, work was on and off with sssd implementation in #22460 which I will see through at some point, but given the limitations with domain vs realm and general expected user workflow for single machines I looked back on `pam-u2f` to see if we can get that to work first due to a better user experience.

I've been looking at implementing `pam-u2f` over on this PR
- https://github.com/cockpit-project/cockpit/pull/22689

Currently blocked on `pam-u2f` not supporting WebAuthn auth assertion due to them not needing it. We are a bit unique in that Cockpit uses PAM modules for authentication by default, which means we need to link a browser WebAuthn to the PAM module. We got info about this through a `pam-u2f` contributor at which told us as much and verified that our assertion will never pass in the current implementation due to missing WebAuthn compliance.
- https://github.com/Yubico/pam-u2f/discussions/381

Me and @Nykseli from SUSE are seeking to make `pam-u2f` compliant with WebAuthn and working together to implement it. I will extend `pam-u2f` to be WebAuthn compliant while @Nykseli  will look over `pam-u2f` dependency `libfido2` to assure it works with WebAuthn authentication assertion

Everything does seem plausible though and initial review of existing `pam-u2f` codebase and W3C spec makes it look like we don't need to change too much to get it to work. No ETA yet though!

**Comment by Venefilyn** [comment]
Just a quick rundown of what needs to happen for WebAuthn to be supported in Cockpit. But in short, Cockpit is currently blocked on the PAM authentication.

## PAM and WebAuthn

While looking around and what exists there is really only one PAM solution that exists which could technically be made to work with WebAuthn. The problem Cockpit has with WebAuthn and PAM is very unique to us as there aren't that many tools out there that link a front-end login directly to PAM - so even `pam-u2f` doesn't have proper WebAuthn support. I have experimented with `pam-u2f` over on https://github.com/cockpit-project/cockpit/pull/22689

Taken from our username/password authentication docs, here is how Cockpit's PAM auth works in general.

```mermaid
sequenceDiagram
    participant User
    participant cockpit-ws
    participant cockpit-session
    participant PAM

    User->>cockpit-ws: GET https://server:9090
    cockpit-ws->>User: 401 + Login page
    User->>cockpit-ws: GET with Authorization: Basic
    cockpit-ws->>cockpit-session: Spawn
    cockpit-session->>cockpit-ws: authorize command with * challenge
    cockpit-ws->>cockpit-session: user/password
    cockpit-session->>PAM: pam_authenticate()
    PAM-->>cockpit-session: Success or conversation
    opt 2FA/password change
        cockpit-session->>cockpit-ws: X-Conversation challenge
        cockpit-ws->>User: Display prompt
        User->>cockpit-ws: Response
        cockpit-ws->>cockpit-session: Forward response
        cockpit-session->>PAM: Continue conversation
    end
    PAM->>cockpit-session: Success
    cockpit-session->>PAM: open_session()
    cockpit-ws->>User: 200 OK for the login page
```

Linking the client with the server requires proper handling of not only the [auth assertion](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API/Attestation_and_Assertion#assertion), and [relying party](https://developer.mozilla.org/en-US/docs/Glossary/Relying_party) (domain/realm), but also [`clientDataJSON`](https://developer.mozilla.org/en-US/docs/Web/API/AuthenticatorResponse/clientDataJSON).

### pam-u2f

This one is the most promising to us still. It handles most of the things we need and can handle our workflow quite well and it does support a manual input mode which is needed for us as we do not have the hardware token available on the server but instead provide data through the client/browser. Since it is a PAM module we get the benefits of not needing any IPA server or custom server implementation. However, it doesn't implement the `clientDataJSON` verification fully, AKA [Limited Verification Algorithm](https://w3c.github.io/webauthn/#clientdatajson-verification). We did get a [potential green light](https://github.com/Yubico/pam-u2f/discussions/381#discussioncomment-15517125) to implement this in `pam-u2f` and both me and @Nykseli have been working on it together (mostly @Nykseli as `pam-u2f` is written in clang and I've been working on my Cockpit passwordless login talk at FOSDEM). Given we don't need to change the codebase too much I'm fairly confident we'll get the feature merged.

What `pam-u2f` was missing in terms of the Limited Verification Algorithm had to do with them using `clientDataHash` for verification and skipping a few steps with domain and challenge verification as it wasn't needed by `pam-u2f`. This really meant that the client assertion could never be verified by `pam-u2f` as the `clientDataHash` would always be different from what `pam-u2f` expects. `clientDataHash` from the specs:

> Let `clientDataHash` be the hash of the serialized client data represented by `clientDataJSON`.

Which means that `pam-u2f` was really just using the hash as a challenge since it didn't care about the [fields type, origin, or crossOrigin](https://www.w3.org/TR/webauthn-3/#client-data). It also makes no real domain verification since phising from what I can tell is a non-issue for terminal-based passkey auth. When we authenticate with the client, the browser creates something like this for the `clientDataJSON`:

```json
{
  "type": "webauthn.get",
  "challenge": "C6cpczNRoqDBP8ZsoqztbQb0W1Sxk_4v3nIpgHymki8",
  "origin": "https://localhost:9091",
  "crossOrigin": false
}
```

So we need verification for `type` and `origin` within `pam-u2f` and that is exactly what @Nykseli is working on in https://github.com/Nykseli/pam-u2f/tree/pam/webauth-option and if this all gets merged it seems likely we will be able to get Cockpit working with passkeys. I've yet to test this code in a VM to see if it all works as expected, hopefully I can get that done before my talk.

### Building our own PAM

Preferably we'd avoid doing this since it means we have more code to maintain and tests to ensure its compliance. But if we cannot get `pam-u2f` to accept the contribution due to maintainence reasons and whatnot there are ways we can get PAM to work. If this happens, hopefully we can get `pam-u2f` to expose a few functions and make it extendable so we can just wrap around that, otherwise it is possible to use the same library it uses - namely `libfido2`. The `libfido2` codebase from what we can tell does support handling WebAuthn's Limited Verification Algorithm and we just need to call the right functions with the right data.

But again, I'd want to avoid this and if we have to do this we'd likely postpone WebAuthn feature even more.

## sssd/FreeIPA and WebAuthn

This would be interesting to add to Cockpit as well, and in fact I think we can do this eventually although likely wouldn't work well due to having to use both web domain and realm for Relying Party (RP) verification. I was trying this out a bit in #22460. 

A while ago sssd and FreeIPA implemented WebAuthn support and it does work quite well. It uses the realm as the RP/origin and generates the challenge and everything correct. With registering a passkey we can utilize [`sssctl`](https://sssd.io/design-pages/passkey_authentication.html#sssctl) and for authentication we likely would need the ability to provide data without relying on a hardware token. I haven't checked if the authentication workflow does work nor have I checked if we can authenticate manually like `pam-u2f` by providing data through stdin or similar.

## Cockpit changes needed

We need to support two authentication workflows, one when we have login with passkeys as-is, and the other with passkeys as a second-step verification. For the workflow itself I'd imagine the first one to be the simplest and is what I'm working on currently. For the auth workflow for discoverable (resident) and non-discoverable (non-resident) credentials I'd look something like this:

```mermaid
sequenceDiagram
    participant Credentials Manager
    actor User
    participant Client
    participant cockpit-ws
    participant cockpit-session

    User->>Client: Go to https://server:9090
    Client->>cockpit-ws: GET https://server:9090
    cockpit-ws->>Client: 401 + Login page
    User->>Client: Click 'Login with passkey'
    Client->>cockpit-ws: GET with Authorization: Passkey
    cockpit-ws->>cockpit-session: Spawn
    cockpit-session->>cockpit-ws: authorize command with * challenge
    cockpit-ws->>cockpit-session: empty content
    create participant PAM
    cockpit-session->>PAM: pam_authenticate()
    PAM-->>cockpit-session: Send PAM challenge
    cockpit-session->>cockpit-ws:X-Conversation challenge
    cockpit-ws->>Client: PAM Challenge
    Client->>Client: Start WebAuthn auth with PAM challenge
    Client->>Credentials Manager: Ask for passkey with Relying Party
    Credentials Manager->>User: Show popup to chose passkey
    User->>Credentials Manager: Choose passkey
    Credentials Manager->>Client: Provide passkey
    Client->>Client: Generate WebAuthn auth assertion
    Client->>cockpit-ws: GET with Authorization: Passkey
    cockpit-ws->>cockpit-session: Forward passkey
    cockpit-session->>PAM: Continue conversation
    PAM->>PAM: Validate challenge
    PAM->>PAM: Validate clientDataJSON
    PAM->>PAM: Validate assertion with stored public key
    PAM->>cockpit-session: Success
    destroy PAM
    cockpit-session->>PAM: open_session()
    cockpit-session->>cockpit-ws: OK
    cockpit-ws->>Client: 200 OK for the login page
```

Or in simpler terms:

```mermaid
sequenceDiagram
    participant Client
    participant Cockpit

    Client->>Cockpit: GET https://server:9090
    Cockpit->>Client: 401 + Login page
    Client->>Cockpit: GET with Authorization: Passkey
    Note over Client,Cockpit: Either an empty username for discoverable keys<br>or pre-filled for non-discoverable keys
    create participant PAM
    Cockpit->>PAM: pam_authenticate()
    PAM-->>Cockpit: Send challenge
    Cockpit->>Client: Forward Challenge
    Note over Client,Cockpit: If non-discoverable this includes<br>accepted list of passkey IDs
    Client->>Client: Get WebAuthn credentials<br>with RP and challenge
    Client->>Cockpit: GET with Authorization: Passkey
    Note over Client,Cockpit: Now includes clientDataJSON, assertion,<br>(and userHandle if discoverable)
    Cockpit->>PAM: Continue Conversation
    PAM->>Cockpit:Success
    destroy PAM
    Cockpit->>PAM: open_session()
    Cockpit->>Client: 200 OK for the login page
```

But we'd need to make this also work with `Authorization: Basic` for when doing 2FA. Which is something I'd imagine requires more design work on `cockpit-session` and is currently unknown for me. Thus for the purposes of implementation needs I'll only focus on authenticating with passkeys as-is and not as a 2FA step.

### `cockpit-session`

With the `cockpit-session` I'm thinking of just extending the existing clang code to support `X-Conversation` response including newlines. Then we need to get the passkey credentials and provide it to `pam-u2f`. 

### Login 

Likely the easiest one, we create a new button `Log in with passkey` that calls `navigator.credentials.get()` and we then formulate a request to Cockpit with `Authorization: Passkey`. Not entirely sure yet of how we will show this button dynamically depending if you have passkeys enabled or not.

### Registering from shell

Similarly to how we add SSH keys at the moment, we can implement another menu item in the navbar's "Session" dropdown for Passkeys if `pam-u2f` is enabled, there we can manage and call `navigator.credentials.create()` and populate the `/etc/u2f_mappings` or user-specific mapping file. We will need some mapping here for what is expected from `pam-u2f`.

<img width="701" height="466" alt="Session dropdown showing style buttons 'default', 'light', and 'dark'. Following that is SSH keys item and below that is a Log out item. A big cursor is shown hovering over SSH keys item and covering parts of the Log out" src="https://github.com/user-attachments/assets/7ee6e33f-c310-493f-a316-91e0e9ec1c2f" />

## Concerns/Thoughts

- `sssd` will likely have issues with the web domain used for registering/authenticating and realm used for the FreeIPA server.
  - This could possibly be mitigated with a current draft for `/.well-known/webauthn` but that also raises the issue regarding having a local realm like "something.local" or "localhost" available for authentication as the browser considers these to be development domains and has less security verification enabled.
- For `pam-u2f`, if the web domain ever changes or if we have multiple domains for the same server (e.g. `cockpit.example.com` and `10.35.30.5`), how can we ensure the registration and auth still works? Currently `pam-u2f` sets the `origin` and `appid` statically through the PAM config, but doesn't allow a list of allowed domains due to the currently limited implementation.
- Likely not an issue as we can configure another token, but would there be issues with authenticating through the browser vs terminal?
- If we need to escalate permissions when already logged in, do we need to do anything special?
- `pam-u2f` also expects some CBOR-encoded data and specific public key fetching from assertion, to what extent do we want to use a CBOR library or do we want to have our own file for it? At the moment I've taken CBOR.js from https://github.com/adricasti/libfido2-test/blob/main/cbor.js

**Comment by JanZerebecki** [comment]
> Likely not an issue as we can configure another token, but would there be issues with authenticating through the browser vs terminal?

The usual way this can be done is login on the terminal (or GUI) displays a link / qr code (may involve a one time code, don't remember, I think is also a variant with Bluetooth) you use somewhere else to authenticate the login request. I have seen a video demonstration of this, claiming its implemented in sssd.

**Comment by JanZerebecki** [comment]
> With registering a passkey we can utilize [sssctl](https://sssd.io/design-pages/passkey_authentication.html#sssctl) and for authentication we likely would need the ability to provide data without relying on a hardware token.

My understanding was that it was all handled by ~~FreeIPA~~ an external IdP (seems just sssd with FreeIPA do not support that, something else needs to be involved, see two comments below) in the webbrowser (though I have not personally tested it), so 1) whether or not it is a hardware token is only relevant to the webbrowser and 2) this isn't relevant to cockpit as it never sees the Webauthn data.

**Comment by JanZerebecki** [comment]
Note that the word passkeys is possibly confusing here, sssd and FreeIPA also support fido2 hardware tokens where the sssd client process talks to it via USB. That has nothing to do with the Webauthn passkey flow, where only the browser ever talks to whatever provides the passkey, possibly a password manager in software.

**Comment by JanZerebecki** [comment]
Ah, found where I had seen one variant of it. *arg* it is using another external idp from freeipa and OAuth 2.0 Device Authorization Grant Flow.

Screenshot, see previous slides for some more info: https://speakerdeck.com/frasertweedale/passwordless-linux-passkey-and-external-idp-support-in-freeipa?slide=30

The screenshot shows a terminal login which prompts you to go to github to authenticate and use a pin to auth the terminal session.

FreeIPA design docs: https://freeipa.readthedocs.io/en/latest/designs/external-idp/external-idp.html

So maybe neither sssd nor FreeIPA support Webauthn?

**Comment by JanZerebecki** [comment]
Note that kanidm claims to support Webauthn directly: https://kanidm.github.io/kanidm/stable/accounts/authentication_and_credentials.html#passkeys

**Comment by JanZerebecki** [comment]
sssd supports Webauthn together with Keycloak (no FreeIPA involved):
* https://sssd.io/docs/idp/idp-introduction.html
* https://keycloak.ch/keycloak-tutorials/tutorial-passkey/

**Comment by Venefilyn** [comment]
@JanZerebecki try to keep it to fewer messages in the future as some get notifications through email 🙏 

Relying inline

---

> > Likely not an issue as we can configure another token, but would there be issues with authenticating through the browser vs terminal?
> 
> The usual way this can be done is login on the terminal (or GUI) displays a link / qr code (may involve a one time code, don't remember, I think is also a variant with Bluetooth) you use somewhere else to authenticate the login request. I have seen a video demonstration of this, claiming its implemented in sssd.

If it is implemented in sssd that would be nice, though if FreeIPA supports it is another matter. But likely one could use a different passkey for browser and for terminal.

`pam-u2f` can likely do the same with different passkeys too


> > With registering a passkey we can utilize [sssctl](https://sssd.io/design-pages/passkey_authentication.html#sssctl) and for authentication we likely would need the ability to provide data without relying on a hardware token.
> 
> My understanding was that it was all handled by ~FreeIPA~ an external IdP (seems just sssd with FreeIPA do not support that, something else needs to be involved, see two comments below) in the webbrowser (though I have not personally tested it), so 1) whether or not it is a hardware token is only relevant to the webbrowser and 2) this isn't relevant to cockpit as it never sees the Webauthn data.

If you've setup FreeIPA to use sssd then you will be able to use passkeys without much issue. You get RP assertion through the realm and you use the hardware tokens directly. [Can see a talk about it here](https://archive.fosdem.org/2024/schedule/event/fosdem-2024-3753-passkey-authentication-the-result/) which goes over registering and authenticating passkeys with FreeIPA and sssd.

But what I meant with my comment was more that I don't know if FreeIPA/sssd can support stdin or other communication means for WebAuthn authentication. If it only supports detecting and communicating with hardware tokens we will have issues with our communication with the browser as no hardware will be detected by FreeIPA/sssd and we have no way of providing the data (and likely no way to get the challenge). So that's why it needs some investigation

> Note that the word passkeys is possibly confusing here, sssd and FreeIPA also support fido2 hardware tokens where the sssd client process talks to it via USB. That has nothing to do with the Webauthn passkey flow, where only the browser ever talks to whatever provides the passkey, possibly a password manager in software.

FIDO2 has overlap with WebAuthn and for basic passkey authentication we can assume a lot of the same things. I.e. validating the assertion, checking the `clientDataJSON`, and matching the user. @ikerexxe who works on sssd can likely go into more detail if you want

> Note that kanidm claims to support Webauthn directly: [kanidm.github.io/kanidm/stable/accounts/authentication_and_credentials.html#passkeys](https://kanidm.github.io/kanidm/stable/accounts/authentication_and_credentials.html#passkeys)

We can't use their PAM module independently of kanidm AFAICT. If we were to support kanidm it would mean a larger effort. Since Cockpit already supports `sssd` we have a much easier time to implement WebAuthn auth and linking it to `sssd`

**Comment by ikerexxe** [comment]
Sorry, this is a long conversation and I don't have the time to read it completely right now, so I'll just try to clarify what I can. Hopefully we'll be able to discuss this topic in person in a few days.

> > The usual way this can be done is login on the terminal (or GUI) displays a link / qr code (may involve a one time code, don't remember, I think is also a variant with Bluetooth) you use somewhere else to authenticate the login request. I have seen a video demonstration of this, claiming its implemented in sssd.
> 
> If it is implemented in sssd that would be nice, though if FreeIPA supports it is another matter. But likely one could use a different passkey for browser and for terminal.

I think you are mixing things here. FreeIPA and SSSD support several authentication mechanisms including passkey and External Identity Providers (EIdP). The link and QR code that you mention come from the EIdP mechanism, not from passkey. I'm not completely sure on how cockpit works and integrates with PAM and SSSD, but if you support this integration you probably already support EIdP.

> But what I meant with my comment was more that I don't know if FreeIPA/sssd can support stdin or other communication means for WebAuthn authentication. If it only supports detecting and communicating with hardware tokens we will have issues with our communication with the browser as no hardware will be detected by FreeIPA/sssd and we have no way of providing the data (and likely no way to get the challenge). So that's why it needs some investigation

The communication between SSSD and passkey uses libfido2. There isn't any browser.

**Comment by Venefilyn** [comment]
> The communication between SSSD and passkey uses libfido2. There isn't any browser.

Yeah my bad, meant from Cockpit itself. `libfido2` does support validating assertions without the hardware token and that is how we are doing it with `pam-u2f`. I.e. Cockpit with WebAuthn provides `clientDataJSON` and assertion to `pam-u2f` which hands it off to `libfido2` for verification.

So I meant investigating whether or not SSSD can authenticate a passkey "manually" without it having hardware access itself, that way we could implement WebAuthn authentication from Cockpit's browser interface to SSSD itself

**Comment by ikerexxe** [comment]
> Yeah my bad, meant from Cockpit itself. `libfido2` does support validating assertions without the hardware token and that is how we are doing it with `pam-u2f`. I.e. Cockpit with WebAuthn provides `clientDataJSON` and assertion to `pam-u2f` which hands it off to `libfido2` for verification.
> 
> So I meant investigating whether or not SSSD can authenticate a passkey "manually" without it having hardware access itself, that way we could implement WebAuthn authentication from Cockpit's browser interface to SSSD itself

Yes, that should be possible, though I'm not certain it would work 'out of the box' without some configuration. SSSD utilizes a separate child process to communicate with libfido2, and this process is already capable of verifying assertion data without requiring a locally attached hardware device.

In fact, this 'offline' verification is a key part of the existing workflow when Kerberos is in use. I'm not exactly sure how to trigger this workflow from SSSD, but it's a matter of integrating it with Cockpit so that it matches.

**Comment by mbof** [comment]
FYI I prototyped an implementation of passkey-based authentication for Cockpit in this project: https://github.com/mbof/pam-paski. It implements a PAM wrapper around the WebAuthn protocol, and injects JS into Cockpit to intercept challenges and communicate them with the passkey. I don't know if the approach is viable (it would likely need some cleanup) or interesting to you, just sharing 😃
